File: src/week4/P004.java
package week4;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class P004 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken()); // 노드(학생) 수
        int m = Integer.parseInt(st.nextToken()); // 간선(키 비교) 수

        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        int[] inDegree = new int[n + 1];

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            // a가 b보다 앞에 서야 함 (a -> b)
            graph.get(a).add(b);
            // b의 진입 차수 증가
            inDegree[b]++;
        }
        Queue<Integer> queue = new LinkedList<>();
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++) {
            if (inDegree[i] == 0) {
                queue.add(i);
            }
        }
        while (!queue.isEmpty()) {
            int s = queue.poll();
            sb.append(s).append(" ");

            for (int adj_s : graph.get(s)) {
                inDegree[adj_s]--;

                if (inDegree[adj_s] == 0) {
                    queue.add(adj_s);
                }
            }
        }
        System.out.println(sb.toString().trim());
    }
}

File: src/week4/P004.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const graph = Array.from({ length: n + 1 }, () => []);
const inDegree = new Array(n + 1).fill(0);

for (let i = 1; i <= m; i++) {
    const [a, b] = input[i].split(' ').map(Number);
    // a -> b (a가 b의 선행)
    graph[a].push(b);
    // b의 진입 차수 증가
    inDegree[b]++;
}

const queue = [];
let head = 0; // shift()를 대체할 포인터
const ans = [];

for (let i = 1; i <= n; i++) {
    if (inDegree[i] === 0) {
        queue.push(i);
    }
}
while (head < queue.length) {
    // shift() 대신 head 포인터로 O(1) 디큐
    const s = queue[head++];
    // 결과 배열에 추가
    ans.push(s);

    // 현재 노드(s)와 연결된 노드들(adj_s)을 순회
    for (const adj_s of graph[s]) {
        // 연결된 노드의 진입 차수 1 감소
        inDegree[adj_s]--;

        // 진입 차수가 0이 되었다면, 큐에 추가
        if (inDegree[adj_s] === 0) {
            queue.push(adj_s);
        }
    }
}
console.log(ans.join(' '));